eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/OnTheFlyFunding.scala

```diff
@@ -0,0 +1,319 @@
+          Behaviors.stopped
+        case WrappedChannelInfo(_, data: DATA_NORMAL) =>
+          nodeParams.db.liquidity.getOnTheFlyFundingPreimage(paymentHash) match {
+            case Some(preimage) =>
+              // We have already received the preimage for that payment, but we probably restarted before removing the
+              // on-the-fly funding proposal from our DB. We must not relay the payment again, otherwise we will pay
+              // the next node twice.
+              cmd.replyTo ! RelaySuccess(channelId, paymentHash, preimage, cmd.status.remainingFees)
+              Behaviors.stopped
+            case None => relay(data)
+          }
+        case WrappedChannelInfo(state, _) =>
+          cmd.replyTo ! RelayFailed(paymentHash, ChannelNotAvailable(state))
+          Behaviors.stopped
+      }
+    }
+
+    private def paymentAlreadyRelayed(paymentHash: ByteVector32, data: DATA_NORMAL): Boolean = {
+      data.commitments.changes.localChanges.all.exists {
+        case add: UpdateAddHtlc => add.paymentHash == paymentHash && add.fundingFee_opt.nonEmpty
+        case _ => false
+      }
+    }
+
+    private def relay(data: DATA_NORMAL): Behavior[Command] = {
+      context.log.debug("relaying {} on-the-fly HTLCs that have been funded", cmd.proposed.size)
+      val htlcMinimum = data.commitments.params.remoteParams.htlcMinimum
+      val cmdAdapter = context.messageAdapter[CommandResponse[CMD_ADD_HTLC]](WrappedCommandResponse)
+      val htlcSettledAdapter = context.messageAdapter[RES_ADD_SETTLED[Origin.Hot, HtlcResult]](WrappedHtlcSettled)
+      cmd.proposed.foldLeft(cmd.status.remainingFees) {
+        case (remainingFees, p) =>
+          // We always set the funding_fee field, even if the fee for this specific HTLC is 0.
+          // This lets us detect that this HTLC is an on-the-fly funded HTLC.
+          val htlcFees = LiquidityAds.FundingFee(remainingFees.min(p.maxFees(htlcMinimum)), cmd.status.txId)
+          val origin = Origin.Hot(htlcSettledAdapter.toClassic, p.upstream)
+          // We only sign at the end of the whole batch.
+          val commit = p.htlc.id == cmd.proposed.last.htlc.id
+          val add = CMD_ADD_HTLC(cmdAdapter.toClassic, p.htlc.amount - htlcFees.amount, paymentHash, p.htlc.expiry, p.htlc.finalPacket, p.htlc.blinding_opt, 1.0, Some(htlcFees), origin, commit)
+          cmd.channel ! add
+          remainingFees - htlcFees.amount
+      }
+      waitForResult(cmd.proposed.size)
+    }
+
+    private def waitForResult(remaining: Int): Behavior[Command] = {
+      Behaviors.receiveMessagePartial {
+        case WrappedCommandResponse(response) => response match {
+          case _: CommandSuccess[_] => Behaviors.same
+          case failure: CommandFailure[_, _] =>
+            cmd.replyTo ! RelayFailed(paymentHash, CannotAddToChannel(failure.t))
+            stopping(remaining - 1)
+        }
+        case WrappedHtlcSettled(settled) =>
```

Comment:
I find the parallel processing of `WrappedCommandResponse`/`WrappedHtlcSettled` a little bit hard to follow, tried a stash-based alternative in 00b1efa5d.